Footsteps effects based on surface type.

fabriciomsantos opened the pull request:
Spawn sounds, decals and Niagara particles based on the surface type defined on the physical material on the mesh. 
The values are defined in a data table to be easily modified.
FALSHitFX struct can be used in other places, for example, bullet holes.

Can be used with https://www.unrealengine.com/marketplace/en-US/product/niagara-footstep-vfx

dyanikoglu commented on Source/ALSV4_CPP/Private/Character/Animation/Notify/ALSAnimNotifyFootstep.cpp:
```diff
@@ -8,29 +8,103 @@
+				
+            for (auto& Data : HitDataTable->GetRowMap())
+            {
+                if (FALSHitFX* HitFX = reinterpret_cast<FALSHitFX*>(Data.Value))
```
Please use Cast() instead.

dyanikoglu commented on Source/ALSV4_CPP/Private/Character/Animation/Notify/ALSAnimNotifyFootstep.cpp:
```diff
@@ -8,29 +8,100 @@
+
+			TArray<FALSHitFX*> HitFXRows;
+			HitDataTable->GetAllRows<FALSHitFX>(FString(),HitFXRows);
+            for (auto& HitFX : HitFXRows)
```
Please fix indentation

dyanikoglu commented:
I'm aiming to test & merge this PR this weekend, thanks for awesome work 👍🏻

fabriciomsantos commented:
There are still some animations that need to be modified to use this new system.
Do you prefer the animations to be modified in this PR or is it better to open another one?

dyanikoglu commented:
![Screenshot 2021-03-18 101147](https://user-images.githubusercontent.com/11363479/111586809-71610e00-87d2-11eb-80d7-4737ac6046d2.png)

I'm currently not able to build this PR after merging into dev branch

fabriciomsantos commented:
This doesn't seem to be a problem caused by this PR, because I didn't modify FALSComponentAndTransform and ALSMathLibrary

This probably happens because of the use of the bEnforceIWYU = true; in ALSV4_CPP.Build.cs and it's solved by removing #include "Library/ALSCharacterStructLibrary.h" from ALSMathLibrary.cpp and adding in ALSMathLibrary.h and removing struct FALSComponentAndTransform; in ALSMathLibrary.h

![1](https://user-images.githubusercontent.com/37518637/111631098-efc9aa00-87ea-11eb-8e4e-8647e75a94ad.png)
![2](https://user-images.githubusercontent.com/37518637/111631110-f22c0400-87ea-11eb-8890-88ebc340213d.png)
![3](https://user-images.githubusercontent.com/37518637/111631121-f3f5c780-87ea-11eb-8ae0-30a518263bce.png)

Erlandys commented:
I would suggest couple improvements:
1. Some animations, like landing has both foot placed to the ground at the same time, which would require for Niagara/Decal to appear at the same moment, so I would suggest to allow specify two separate configurations in same notify for both feet.
2. Some animations has Volume/Pitch multipliers applied, without enabling 'bOverrideMaskCurve', since it then multiplies value from notify with curve value, in your implementation, to change Volume/Pitch multiplier, requires to enable 'bOverrideMaskCurve', which will override value, not multiply it.
3. I'd suggest to split decal info in datatable to two separate 'Left' and 'Right' foot decals, maybe someone will have to separate decals for different feet, without mirroring same decal?
4. I think it could be good, to have some default footstep value from datatable, if no physics surface were returned by hit, but hit were successful, to not have silent step, if some parts were missing physics surface?
5. Also maybe, if shipping, perform `World->LineTraceSingleByChannel` instead of using kismet library?
6. Maybe better design pattern would be for early returns, instead of putting base code to very deep scope, this could be achieved, by using `FindByPredicate` from `HitFXRows` instead of iterating it and also returning if hit failed.
7. Also, I think location offsets for Niagara particles/Decals should be calculated with transforms, not just by simple addition. This will result in more correct offset behavior, since if you add offset by x/y, while character rotates, rotation will not be taken into account and result position will always differ:
```
TransformA(FootRotation, FootLocation);
TransformB(OffsetRotation, OffsetLocation);
ResultTransform = TransformB * TransformA;
```

If necessary I can post changes I've made myself (already made changes which I've pointed).

Sorry for inconvenience and thanks for awesome addition!

fabriciomsantos commented:
Thanks for the suggestions!

1. I want this to be a generic implementation that can be used with characters with any number of feet (such as animals), to have both feet at the same time, you just need to add another notify on another notify track.
2. Now it's fixed.
3. In this case, I think it would be better to create another data table.
4. All collisions already have a default physics material with the default surface type.
5. `UKismetSystemLibrary::LineTraceSingle` already use `World->LineTraceSingleByChannel`, the only difference is some additional checks and a option to debug.
6. Implemented.
7. Fixed using the owner transform.

dyanikoglu commented:
Merged with https://github.com/dyanikoglu/ALS-Community/commit/eebec91245aea8a23e9fe8fea763178ae344458f
